### DISCUSSION TITLE
Stress test WPF data grid - how to make it lag and measure it?

### DISCUSSION THREAD

**Discussion by maxima120** [dotnet/wpf#8493] [Q&A]
I have a .NET 7 WPF app with high volume of updates per second in a data grid (currently devexpress) with 100x100k col/rows.
I would like to stress test it before we go to production. Also would like to run the same with different data grids to compare (telerik and syncfusion)..

One idea is to saturate the window queue with batch of messages, store the timestamp of the last message then find timepstamp of the last UI update. Not sure how to achieve it. If I feed UI from 4 background threads without sleep or wait - non-stop, and use `Dispatcher.BeginInvoke` on the window - will i be able to measure actual delay?

Basically i want to make it lag **and** be able to measure that lag. Anyone did this? Could you please give a hint?

Here is what I came up (it seems meaningful but would love to hear other people point of view):
[rows = 10,000; columns = 100;  workers = 1; num = 1mil]
[`d` - parent window dispatcher, `GetStats` - called manually by a button click]
```
        public long startTimestamp = 0L;
        public long lastWorkerTimestamp = 0L;
        public long lastDispatcherTimestamp = 0L;
        void Worker(int n)
        {
            var rnd = new Random((int)DateTime.Now.Ticks);

            startTimestamp = Stopwatch.GetTimestamp();

            int i = 0;
            while (!stopFlag && i < num)
            {
                var row = rnd.Next(0, rows);
                var col = rnd.Next(0, cols);
                var val = rnd.NextDouble();

                locker.EnterWriteLock();
                hits.Add(DateTime.Now.Ticks);
                locker.ExitWriteLock();

                lastWorkerTimestamp = Stopwatch.GetTimestamp();

                d.BeginInvoke(() =>
                {
                    data[row].SetValue(col, val);
                    // not ideal but should be consistent across all the tests                   
                    lastDispatcherTimestamp = Stopwatch.GetTimestamp();
                });

                i++;
            }
        }


...
        public string GetStats()
        {
            long ts1 = 0L, ts2 = 0L;
            Interlocked.Exchange(ref ts1, lastWorkerTimestamp);
            Interlocked.Exchange(ref ts2, lastDispatcherTimestamp);

            var wall = (double)(lastDispatcherTimestamp - startTimestamp) / Stopwatch.Frequency;
            var value = $"Wallclock: {wall:0.000}s, Lag: {(double)(ts2 - ts1) / Stopwatch.Frequency:0.000}s";

            return value;
        }
```

**Comment by lindexi** [comment]
It's challenging for me to evaluate your measurement method, but I believe that having a more comprehensive understanding of the WPF framework would be beneficial.

**Reply by maxima120** [reply]
if you think that I am doing something stupid - just offer an alternative solution. How do you overload UI thread and measure lag - ie time between when last update was pushed from DAL and time the last update rendered in the UI ?
